README.md
# Laravel Queue Pool

[![Latest Version on Packagist](https://img.shields.io/packagist/v/wanghanlin/laravel-queue-pool.svg?style=flat-square)](https://packagist.org/packages/wanghanlin/laravel-queue-pool)
[![Build Status](https://img.shields.io/travis/wanghanlin/laravel-queue-pool/master.svg?style=flat-square)](https://travis-ci.org/wanghanlin/laravel-queue-pool)
[![Quality Score](https://img.shields.io/scrutinizer/g/wanghanlin/laravel-queue-pool.svg?style=flat-square)](https://scrutinizer-ci.com/g/wanghanlin/laravel-queue-pool)
[![Total Downloads](https://img.shields.io/packagist/dt/wanghanlin/laravel-queue-pool.svg?style=flat-square)](https://packagist.org/packages/wanghanlin/laravel-queue-pool)

This package provide an artisan command `queue:pool` to start multiple queue worker and supervise them.

## Installation

You can install the package via composer:

```bash
composer require wanghanlin/laravel-queue-pool
```

## Usage

```bash
php artisan queue:pool --workers=3
php artisan queue:pool connection --workers=3 --queue=default --delay=0 --force --memory=128 --sleep=3 --timeout=60 --tries=0
```

### Testing

``` bash
composer test
```

### Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

### Security

If you discover any security related issues, please email 605231181@qq.com instead of using the issue tracker.

## Credits

- [Hanlin Wang](https://github.com/wanghanlin)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `laravel-queue-pool` will be documented in this file

## 0.0.1 - 2018-02-13

- initial release


tests/QueuePoolTest.php
<?php

namespace Wanghanlin\QueuePool\Tests;

use Mockery as m;
use PHPUnit\Framework\TestCase;
use Wanghanlin\QueuePool\QueuePool;
use Wanghanlin\QueuePool\QueuePoolOption;
use Mockery\Adapter\Phpunit\MockeryPHPUnitIntegration;

class QueuePoolTest extends TestCase
{
    use MockeryPHPUnitIntegration;

    public function testRunProcessCallsProcess()
    {
        $process1 = m::mock('Symfony\Component\Process\Process')->makePartial();
        $process1->shouldReceive('start')->once();

        $process2 = m::mock('Symfony\Component\Process\Process')->makePartial();
        $process2->shouldReceive('start')->once();

        $pool = m::mock('Wanghanlin\QueuePool\QueuePool')->makePartial();

        $pool->shouldReceive('getProcesses')->once()->andReturn([$process1, $process2]);
        $pool->shouldReceive('memoryExceeded')->once()->with(1)->andReturn(false);

        $pool->runProcesses(1);
    }

    public function testPoolCanRestartWorker()
    {
        $process = m::mock('Symfony\Component\Process\Process')->makePartial();

        $process->shouldReceive('start')->once();
        $process->shouldReceive('isRunning')->twice()->andReturn(true, false);

        $pool = m::mock('Wanghanlin\QueuePool\QueuePool')->makePartial();

        $pool->shouldReceive('getProcesses')->twice()->andReturn([$process]);
        $pool->shouldReceive('memoryExceeded')->twice()->with(1)->andReturn(false);

        $pool->runProcesses(1);
        $pool->runProcesses(1);
    }

    public function testPoolStopsWhenMemoryIsExceeded()
    {
        $process = m::mock('Symfony\Component\Process\Process')->makePartial();

        $process->shouldReceive('start')->once();

        $pool = m::mock('Wanghanlin\QueuePool\QueuePool')->makePartial();

        $pool->shouldReceive('getProcesses')->once()->andReturn([$process]);
        $pool->shouldReceive('memoryExceeded')->once()->with(1)->andReturn(true);
        $pool->shouldReceive('stop')->once();

        $pool->runProcesses(1);
    }

    public function testMakeProcessCorrectlyFormatsCommandLine()
    {
        $pool = new QueuePool(__DIR__);

        $options = new QueuePoolOption();
        $options->workers = 1;
        $options->delay = 1;
        $options->memory = 2;
        $options->timeout = 3;

        $processes = $pool->makeProcesses('connection', 'queue', $options);

        $escape = '\\' === DIRECTORY_SEPARATOR ? '"' : '\'';

        $this->assertCount(1, $processes);

        $process = array_first($processes);

        $this->assertInstanceOf('Symfony\Component\Process\Process', $process);

        $this->assertEquals(__DIR__, $process->getWorkingDirectory());
        $this->assertEquals(3, $process->getTimeout());
        $this->assertEquals($escape.PHP_BINARY.$escape." {$escape}artisan{$escape} queue:work {$escape}connection{$escape} --queue={$escape}queue{$escape} --delay=1 --memory=2 --sleep=3 --tries=0", $process->getCommandLine());
    }
}


src/QueuePool.php
<?php

namespace Wanghanlin\QueuePool;

use Illuminate\Support\ProcessUtils;
use Symfony\Component\Process\Process;
use Symfony\Component\Process\PhpExecutableFinder;

class QueuePool
{
    /**
     * The command working path.
     *
     * @var string
     */
    protected $commandPath;

    /**
     * The environment the workers should run under.
     *
     * @var string
     */
    protected $environment;

    /**
     * The worker processes that started.
     *
     * @var array
     */
    protected $processes;

    /**
     * The amount of seconds to wait before polling the queue.
     *
     * @var int
     */
    protected $sleep = 3;

    /**
     * The amount of times to try a job before logging it failed.
     *
     * @var int
     */
    protected $maxTries = 0;

    /**
     * The queue worker command line.
     *
     * @var string
     */
    protected $workerCommand;

    /**
     * The output handler callback.
     *
     * @var \Closure|null
     */
    protected $outputHandler;

    /**
     * Create a new queue listener.
     *
     * @param  string  $commandPath
     * @return void
     */
    public function __construct($commandPath)
    {
        $this->commandPath = $commandPath;
        $this->workerCommand = $this->buildCommandTemplate();
    }

    /**
     * Build the environment specific worker command.
     *
     * @return string
     */
    protected function buildCommandTemplate()
    {
        $command = 'queue:work %s --queue=%s --delay=%s --memory=%s --sleep=%s --tries=%s';

        return "{$this->phpBinary()} {$this->artisanBinary()} {$command}";
    }

    /**
     * Get the PHP binary.
     *
     * @return string
     */
    protected function phpBinary()
    {
        return ProcessUtils::escapeArgument(
            (new PhpExecutableFinder)->find(false)
        );
    }

    /**
     * Get the Artisan binary.
     *
     * @return string
     */
    protected function artisanBinary()
    {
        return defined('ARTISAN_BINARY')
            ? ProcessUtils::escapeArgument(ARTISAN_BINARY)
            : ProcessUtils::escapeArgument('artisan');
    }

    /**
     * Start workers.
     *
     * @param  string  $connection
     * @param  string  $queue
     * @param  QueuePoolOption  $options
     * @return void
     */
    public function start($connection, $queue, QueuePoolOption $options)
    {
        $processes = $this->makeProcesses($connection, $queue, $options);

        $this->setProcesses($processes);

        while (true) {
            $this->runProcesses($options->memory);
        }
    }

    /**
     * Create an array of Symfony processes.
     *
     * @param $connection
     * @param $queue
     * @param QueuePoolOption $options
     * @return array
     */
    public function makeProcesses($connection, $queue, QueuePoolOption $options)
    {
        $processes = [];

        foreach (range(1, $options->workers) as $key) {
            $processes[$key] = $this->makeProcess($connection, $queue, $options);
        }

        return $processes;
    }

    /**
     * Create a new Symfony process for the worker.
     *
     * @param  string  $connection
     * @param  string  $queue
     * @param  QueuePoolOption  $options
     * @return \Symfony\Component\Process\Process
     */
    public function makeProcess($connection, $queue, QueuePoolOption $options)
    {
        $command = $this->workerCommand;

        // If the environment is set, we will append it to the command string so the
        // workers will run under the specified environment. Otherwise, they will
        // just run under the production environment which is not always right.
        if (isset($options->environment)) {
            $command = $this->addEnvironment($command, $options);
        }

        // Next, we will just format out the worker commands with all of the various
        // options available for the command. This will produce the final command
        // line that we will pass into a Symfony process object for processing.
        $command = $this->formatCommand(
            $command, $connection, $queue, $options
        );

        return new Process(
            $command, $this->commandPath, null, null, $options->timeout
        );
    }

    /**
     * Add the environment option to the given command.
     *
     * @param  string  $command
     * @param  QueuePoolOption  $options
     * @return string
     */
    protected function addEnvironment($command, QueuePoolOption $options)
    {
        return $command.' --env='.ProcessUtils::escapeArgument($options->environment);
    }

    /**
     * Format the given command with the listener options.
     *
     * @param $command
     * @param $connection
     * @param $queue
     * @param QueuePoolOption $options
     * @return string
     */
    protected function formatCommand($command, $connection, $queue, QueuePoolOption $options)
    {
        return sprintf(
            $command,
            ProcessUtils::escapeArgument($connection),
            ProcessUtils::escapeArgument($queue),
            $options->delay, $options->memory,
            $options->sleep, $options->maxTries
        );
    }

    /**
     * Run worker processes.
     *
     * @param  int  $memory
     * @return void
     */
    public function runProcesses($memory)
    {
        $processes = $this->getProcesses();

        array_walk($processes, function ($process, $key) {
            if (! $process->isRunning()) {
                $process->start(function ($type, $line) use ($key) {
                    $line = "[Worker $key]: $line";
                    $this->handleWorkerOutput($type, $line);
                });
            }
        });

        // Once we have run the job we'll go check if the memory limit has been exceeded
        // for the script. If it has, we will kill this script so the process manager
        // will restart this with a clean slate of memory automatically on exiting.
        if ($this->memoryExceeded($memory)) {
            $this->stop();
        }
    }

    /**
     * Get processes.
     *
     * @return array
     */
    public function getProcesses()
    {
        return $this->processes;
    }

    /**
     * Set processes.
     *
     * @param array $processes
     */
    public function setProcesses($processes)
    {
        $this->processes = $processes;
    }

    /**
     * Handle output from the worker process.
     *
     * @param  int  $type
     * @param  string  $line
     * @return void
     */
    protected function handleWorkerOutput($type, $line)
    {
        if (isset($this->outputHandler)) {
            call_user_func($this->outputHandler, $type, $line);
        }
    }

    /**
     * Determine if the memory limit has been exceeded.
     *
     * @param  int  $memoryLimit
     * @return bool
     */
    public function memoryExceeded($memoryLimit)
    {
        return (memory_get_usage() / 1024 / 1024) >= $memoryLimit;
    }

    /**
     * Stop listening and bail out of the script.
     *
     * @return void
     */
    public function stop()
    {
        die;
    }

    /**
     * Set the output handler callback.
     *
     * @param  \Closure  $outputHandler
     * @return void
     */
    public function setOutputHandler(\Closure $outputHandler)
    {
        $this->outputHandler = $outputHandler;
    }
}


src/QueuePoolOption.php
<?php

namespace Wanghanlin\QueuePool;

use Illuminate\Queue\WorkerOptions;

class QueuePoolOption extends WorkerOptions
{
    /**
     * The amount of pool worker need to start.
     *
     * @var int
     */
    public $workers;

    /**
     * The environment the pool worker should run in.
     *
     * @var string
     */
    public $environment;

    /**
     * Create a new pool options instance.
     *
     * @param  int  $workers
     * @param  string $environment
     * @param  int  $delay
     * @param  int  $memory
     * @param  int  $timeout
     * @param  int  $sleep
     * @param  int  $maxTries
     * @param  bool  $force
     * @return void
     */
    public function __construct($workers = 1, $environment = null, $delay = 0, $memory = 128, $timeout = 60, $sleep = 3, $maxTries = 0, $force = false)
    {
        $this->workers = $workers;
        $this->environment = $environment;
        parent::__construct($delay, $memory, $timeout, $sleep, $maxTries, $force);
    }
}


src/QueuePoolServiceProvider.php
<?php

namespace Wanghanlin\QueuePool;

use Illuminate\Support\ServiceProvider;

class QueuePoolServiceProvider extends ServiceProvider
{
    /**
     * Bootstrap the application services.
     */
    public function boot()
    {
        $this->app->singleton('queue.pool', function () {
            return new QueuePool($this->app->basePath());
        });

        $this->app->singleton('command.queue.pool', function ($app) {
            return new QueuePoolCommand($app['queue.pool']);
        });

        if ($this->app->runningInConsole()) {
            $this->commands([
                'command.queue.pool',
            ]);
        }
    }

    /**
     * Register the application services.
     */
    public function register()
    {
        //
    }

    /**
     * Get the services provided by the provider.
     *
     * @return array
     */
    public function provides()
    {
        return [
            'queue.pool', 'command.queue.pool',
        ];
    }
}


src/QueuePoolCommand.php
<?php

namespace Wanghanlin\QueuePool;

use Illuminate\Console\Command;

class QueuePoolCommand extends Command
{
    /**
     * The console command name.
     *
     * @var string
     */
    protected $signature = 'queue:pool
                            {connection? : The name of the queue connection to work}
                            {--workers= : The amount of the workers to start}
                            {--queue= : The names of the queues to work}
                            {--delay=0 : The number of seconds to delay failed jobs}
                            {--force : Force the worker to run even in maintenance mode}
                            {--memory=128 : The memory limit in megabytes}
                            {--sleep=3 : Number of seconds to sleep when no job is available}
                            {--timeout=60 : The number of seconds a child process can run}
                            {--tries=0 : Number of times to attempt a job before logging it failed}';

    /**
     * The console command description.
     *
     * @var string
     */
    protected $description = 'Start a pool of workers to process the queue';

    /**
     * The queue pool instance.
     *
     * @var QueuePool
     */
    protected $pool;

    /**
     * Create a new queue work command.
     *
     * @param  QueuePool  $pool
     * @return void
     */
    public function __construct(QueuePool $pool)
    {
        parent::__construct();
        $this->setOutputHandler($this->pool = $pool);
    }

    /**
     * Execute the console command.
     *
     * @return void
     */
    public function handle()
    {
        // We need to get the right queue for the connection which is set in the queue
        // configuration file for the application. We will pull it based on the set
        // connection being run for the queue operation currently being executed.
        $queue = $this->getQueue(
            $connection = $this->input->getArgument('connection')
        );
        $this->pool->start(
            $connection, $queue, $this->gatherOptions()
        );
    }

    /**
     * Get the name of the queue connection to listen on.
     *
     * @param  string  $connection
     * @return string
     */
    protected function getQueue($connection)
    {
        $connection = $connection ?: $this->laravel['config']['queue.default'];

        return $this->input->getOption('queue') ?: $this->laravel['config']->get(
            "queue.connections.{$connection}.queue", 'default'
        );
    }

    /**
     * Get the listener options for the command.
     *
     * @return QueuePoolOption
     */
    protected function gatherOptions()
    {
        return new QueuePoolOption(
            $this->option('workers'), $this->option('env'),
            $this->option('delay'), $this->option('memory'),
            $this->option('timeout'), $this->option('sleep'),
            $this->option('tries'), $this->option('force')
        );
    }

    /**
     * Set the options on the queue listener.
     *
     * @param  QueuePool  $pool
     * @return void
     */
    protected function setOutputHandler(QueuePool $pool)
    {
        $pool->setOutputHandler(function ($type, $line) {
            $this->output->write($line);
        });
    }
}
